rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,141 @@
+# String interpolation
+
+## Summary
+
+New string interpolation syntax.
+
+## Motivation
+
+The problems with `string.format` are many.
+
+1. Must be exact about the types and its corresponding value.
+2. Using `%d` is the idiomatic default for most people, but this loses precision.
+   * `%d` casts the number into `long long`, which has a lower max value than `double` and does not support decimals.
+   * `%f` by default will format to the millionths, e.g. `5.5` is `5.500000`.
+   * `%g` by default will format up to the hundred thousandths, e.g. `5.5` is `5.5` and `5.5312389` is `5.53123`. It will also convert the number to scientific notation when it encounters a number equal to or greater than 10^6.
+   * To not lose too much precision, you need to use `%s`, but even so the type checker assumes you actually wanted strings.
+3. No support for `boolean`. You must use `%s` **and** call `tostring`.
+4. No support for values implementing the `__tostring` metamethod.
+5. Using `%` is in itself a dangerous operation within `string.format`.
+   * `"Your health is %d% so you need to heal up."` causes a runtime error because `% so` is actually parsed as `(%s)o` and now requires a corresponding string.
+6. Having to use parentheses around string literals just to call a method of it.
+
+## Design
+
+To fix all of those issues, we need to do a few things.
+
+1. A new string interpolation expression (fixes #5, #6)
+2. Extend `string.format` to accept values of arbitrary types (fixes #1, #2, #3, #4)
+
+Because we care about backward compatibility, we need some new syntax in order to not change the meaning of existing strings. There are a few components of this new expression:
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
+   * Restriction: the string interpolation literal must have at least one value to interpolate. We do not need 3 ways to express a single line string literal.
+   * The pairs must be on the same line (unless a `\` escapes the newline) but expressions needn't be on the same line.
+2. An expression between the braces. This is the value that will be interpolated into the string.
+3. Formatting specification may follow the expression, delimited by an unambiguous character.
+   * Restriction: the formatting specification must be constant at parse time.
+   * The syntax for formatting specification and a delimiter is undefined in this RFC. A future RFC will implement this.
```

Comment:
Just to clarify, this is left for a future extension, correct? So this is saying that this RFC doesn't permit formatting specification as part of interpolated strings, but a future RFC may amend that?

Reply:
That's correct.

Reply:
Should I change the wording a bit? I can actually see it being a little ambiguous in that it's "left to the implementor," which is not intended.

Reply:
I think we should be more explicit yeah, smth like "An interpolated expression does not have a formatting specification and is formatted using the default rules equivalent to `tostring` call; a future extension may be introduced to extend the syntax with an optional specification".

Reply:
Okay, I updated that section. I also noticed that it didn't quite say that it would use `%*` by default, so added that in to be explicit.

Reply:
Hey! I have a question! I'm a Roblox developer. So do you guys work at Roblox, or do you make Studio mods or what?! 

- Deebest2010 (Roblox)